File: puslefill/pulsefill/apps/dashboard-web/lib/operator-primary-action.ts
import type { ActionQueueItem } from "@/types/action-queue";
import type { OperatorSlotsListItem } from "@/types/operator-slots-list";

export type DerivedOperatorPrimaryAction =
  | { kind: "send_offers"; label: "Send offers" }
  | { kind: "retry_offers"; label: "Retry" }
  | { kind: "confirm_booking"; label: "Confirm"; claimId: string }
  | null;

/** Slots list + shared status/claim rules. */
export function deriveOperatorPrimaryAction(input: {
  status?: string | null;
  winningClaimId?: string | null;
}): DerivedOperatorPrimaryAction {
  const status = (input.status || "").toLowerCase();

  if (status === "open") {
    return { kind: "send_offers", label: "Send offers" };
  }

  if (status === "offered") {
    return { kind: "retry_offers", label: "Retry" };
  }

  if (status === "claimed" && input.winningClaimId) {
    return {
      kind: "confirm_booking",
      label: "Confirm",
      claimId: input.winningClaimId,
    };
  }

  return null;
}

export function deriveOperatorPrimaryActionFromSlot(slot: OperatorSlotsListItem): DerivedOperatorPrimaryAction {
  return deriveOperatorPrimaryAction({
    status: slot.status,
    winningClaimId: slot.winning_claim?.id ?? null,
  });
}

/**
 * Queue rows: respect API action ordering (e.g. inspect_logs stays non-inline).
 * Uses slot_status + claim_id for confirm / send / retry when appropriate.
 */
export function deriveQueueInlinePrimaryAction(item: ActionQueueItem): DerivedOperatorPrimaryAction {
  const first = item.actions[0];
  if (first === "inspect_logs") {
    return null;
  }
  if (first === "view_slot") {
    return null;
  }

  const status = (item.slot_status || "").toLowerCase();

  if (first === "confirm_booking") {
    if (status === "claimed" && item.claim_id) {
      return { kind: "confirm_booking", label: "Confirm", claimId: item.claim_id };
    }
    return null;
  }

  if (first === "retry_offers") {
    if (status === "open") {
      return { kind: "send_offers", label: "Send offers" };
    }
    if (status === "offered") {
      return { kind: "retry_offers", label: "Retry" };
    }
    return null;
  }

  return null;
}

File: puslefill/pulsefill/ios/PulseFill/PulseFill/Features/Operator/OperatorPrimaryAction.swift
import Foundation

enum OperatorInlineActionKind: Equatable {
    case confirmBooking
    case sendOffers
    case retryOffers
}

struct OperatorPrimaryAction: Equatable {
    let kind: OperatorInlineActionKind
    let label: String
    let claimId: String?
}

enum OperatorPrimaryActionDeriver {
    static func forSlot(status: String, winningClaimId: String?) -> OperatorPrimaryAction? {
        switch status.lowercased() {
        case "open":
            return OperatorPrimaryAction(kind: .sendOffers, label: "Send offers", claimId: nil)
        case "offered":
            return OperatorPrimaryAction(kind: .retryOffers, label: "Retry", claimId: nil)
        case "claimed":
            guard let winningClaimId else { return nil }
            return OperatorPrimaryAction(kind: .confirmBooking, label: "Confirm", claimId: winningClaimId)
        default:
            return nil
        }
    }

    /// Mirrors web `deriveQueueInlinePrimaryAction`: first `actions` entry wins; `view_slot` / `inspect_logs` → no inline CTA.
    static func queueInline(from item: OperatorActionQueueItem) -> OperatorPrimaryAction? {
        guard let first = item.actions.first else { return nil }

        if first == .inspectLogs || first == .viewSlot || first == .openSlot {
            return nil
        }

        let status = (item.slotStatus ?? "").lowercased()

        if first == .confirmBooking {
            guard status == "claimed", let claimId = item.claimId else { return nil }
            return OperatorPrimaryAction(kind: .confirmBooking, label: "Confirm", claimId: claimId)
        }

        if first == .retryOffers {
            if status == "open" {
                return OperatorPrimaryAction(kind: .sendOffers, label: "Send offers", claimId: nil)
            }
            if status == "offered" {
                return OperatorPrimaryAction(kind: .retryOffers, label: "Retry", claimId: nil)
            }
            return nil
        }

        return nil
    }
}
